runtime/tr.source/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -1294,15 +1294,15 @@ VMnonNullSrcWrtBarCardCheckEvaluator(
       TR::InstOpCode::Mnemonic opSubtractReg = TR::InstOpCode::getSubstractRegOpCode();
       TR::InstOpCode::Mnemonic opSubtract = TR::InstOpCode::getSubstractOpCode();
       TR::InstOpCode::Mnemonic opCmpLog = TR::InstOpCode::getCmpLogicalOpCode();
-      uintptrj_t heapSize = (uintptrj_t) comp->getOptions()->getHeapSizeForBarrierRange0();
-      uintptrj_t heapBase = (uintptrj_t) comp->getOptions()->getHeapBaseForBarrierRange0();
+      uint32_t heapSize =  comp->getOptions()->getHeapSizeForBarrierRange0();
+      uint32_t heapBase =  comp->getOptions()->getHeapBaseForBarrierRange0();
```

Comment:
This does not look correct. `uint32_t` is 32-bits. The return value of `getHeapSizeForBarrierRange0` on a 64-bit target will be 64-bit. Is this not a truncation operation?

Reply:
I do not believe my change here causes a functional change. Let me explain. We have 3 variables of interest:
`heapSize`
`heapBase`
`heapSum`

They are only used in:
`generateRILInstruction`
`generateS390CompareAndBranchInstruction`

Before my changes `generateRILInstruction` converted the `uintptrj_t` immediate to an `int32_t` in the `S390Instruction` constructor, so all my change does here is move that conversion further "forward".

`generateS390CompareAndBranchInstruction` takes an `int32_t` as well, so again I'm just moving the conversion "forward".

I agree this seems wrong, but perhaps this is a bug in the existing code? Do you have any suggestions?

Reply:
Looks like `heapSize` and `heapBase` are only used withing the `if (constantHeapCase)` condition.
This condition is defined as:

```
bool constantHeapCase = ((!comp->compileRelocatableCode()) && isConstantHeapBase && isConstantHeapSize && shiftAmount == 0 && (!is64Bit || TR::Compiler->om.generateCompressedObjectHeaders()));
```

Noteworthy here is `shiftAmount == 0` and the subsequent dual check for `!is64Bit`. This means we only enter this `if` condition if our heap fits within 32-bit. 

The only use of `generateRIL` is within the `if` condition so it would be incorrect to change the type of `heapSize` and `heapBase` to `uint32_t` since the variables are declared outside of the `if` condition.

We seem to have two choices (unless there are more?):

1. Lower the definition of `heapBase` and `heapSize` to be inside of `if (constantHeapCase)`, at which point their values are guaranteed to be 32-bit (we should probably note this via a comment)
2. Keep the definitions where they are, revert the type back to `uintptrj_t` and excplicitly cast to `uint32_t` in the `if` where the variables are being used within `generateRILInstruction`

Do we agree?

Reply:
Agreed. Going with option 1) with a comment.